PR: Add an "Fit plots to window" option to the plots pane

gepcel opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
* [x] Included a screenshot (if affecting the UI)

<!--- Explain what you've done and why --->


While using the plotting plugin, I found it annoying to have to zoom in and out the image frequently when testing different sizes of image. Feels like an `Auto fit plotting` option is necessary.

So I tried adding one, like (default to False, aka, unchecked):
![option](https://user-images.githubusercontent.com/4179106/55171564-8b535180-51b3-11e9-8875-c676aa4b3505.png)

An example of the original behavior when dealing with different sizes of images (with the auto_fit_plotting option set to false):
![false](https://user-images.githubusercontent.com/4179106/55171307-0d8f4600-51b3-11e9-8867-a90bfa5f34ef.gif)

Compared to the new behavior when set to true:
![true](https://user-images.githubusercontent.com/4179106/55171399-40d1d500-51b3-11e9-83fe-d5089f3f8455.gif)

Example code:
```python
from matplotlib.pyplot import figure, text
def test_plot(w, h):
    figure(figsize=(w, h))
    text(.5, .5, f'figsize={w}, {h}', ha='center', fontsize=2.5*w)
    
test_plot(4, 3)
test_plot(15, 7)
test_plot(8, 12)
```


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: gepcel

<!--- Thanks for your help making Spyder better for everyone! --->

goanpeca commented:
@gepcel thanks for working on this. Looks pretty useful :-)

@jnsebgosselin could you chime in with your thoughts?

jnsebgosselin commented:
@goanpeca Sure, I'll take a look at it as soon as I can.

CAM-Gerlach commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -290,6 +307,12 @@ def show_fig_outline_in_viewer(self, state):
         self.option_changed('show_plot_outline', state)
 
+    def change_auto_fit_plotting(self, state):
+        """Change the auto_fit_plotting option and scale images"""
```
Per PEP 257, all docstrings must end with periods.

CAM-Gerlach commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -456,10 +481,26 @@ def zoom_out(self):
+        # Scale the image to fit figcanvas size while respect the ratio
+        else:
+            size = self.size()
+            width = size.width() - 15  # How to get the actual figcanvas size?
```
@jnsebgosselin ?

ccordoba12 replied:
@jnsebgosselin, I think we need your help to clarify this. Otherwise, @CAM-Gerlach has done a great job reviewing this PR.

CAM-Gerlach replied:
There's a similar comment in the tests, too, that was added after I reviewed FYI.

gepcel replied:
I think I have made this right, also with the test.

CAM-Gerlach commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -229,7 +237,16 @@ def setup_option_actions(self, mute_inline_plotting, show_plot_outline):
 
-        self.actions = [self.mute_inline_action, self.show_plot_outline_action]
+        self.auto_fit_action = create_action(
+            self, _("Auto fit plotting"),
```
``'Fit plots to window'``

CAM-Gerlach commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -229,7 +237,16 @@ def setup_option_actions(self, mute_inline_plotting, show_plot_outline):
-        self.actions = [self.mute_inline_action, self.show_plot_outline_action]
+        self.auto_fit_action = create_action(
+            self, _("Auto fit plotting"),
+            tip=_("Auto fit plotting in the widget."),
```
``'Automatically fit plots to Plot pane size'``

jnsebgosselin commented:
@goanpeca I'll let @CAM-Gerlach handle this, thanks.

gepcel commented:
@CAM-Gerlach I've updated some commits. Now:

- [x] Affirmed the DCO.
- [x] Added a test. I do need guidance on this, added some code for a start.
- [x] The option is default now.
- [ ]   Shouldn't we make it a persistent config setting so we don't have to set it every time
- [x] Changes based on code comments.

As to the "persistent config setting" suggestion, I'm not sure I understand you correctly, now:
1. After setting `Fit plots to windows`  Checked/Unchecked, the state will be remembered after closing and reopenning spyder, until the next change or until start spyder with `--safe-mode`. 
2. The `Fit plots to windows` state change is not `one plot specific`, it applies to all existing and all after comming plots in the

CAM-Gerlach commented:
> After setting Fit plots to windows Checked/Unchecked, the state will be remembered after closing and reopenning spyder, until the next change or until start spyder with --safe-mode.

This is what I meant, thanks. I thought it wasn't but it was probably because I was switching in and out of ``--safe-mode``, Windows and Linux and lost track.

CAM-Gerlach commented:
Other than your call-out about proper plot window width that perhaps @jnsebgosselin can provide insight on, looks like you addressed all my suggestions, thanks!

pep8speaks commented:
Hello @gepcel! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`spyder/plugins/plots/widgets/tests/test_plots_widgets.py`](https://github.com/spyder-ide/spyder/blob/70500695588784820b78ee8efa87e861d316052c/spyder/plugins/plots/widgets/tests/test_plots_widgets.py):

> [Line 382:1](https://github.com/spyder-ide/spyder/blob/70500695588784820b78ee8efa87e861d316052c/spyder/plugins/plots/widgets/tests/test_plots_widgets.py#L382): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

gepcel commented:
I think I have figured out the accurate plot size when scaling image.

jnsebgosselin commented:
I'll try to take a look at this PR a review it before the end of the week if it is possible to wait a little before merging this.

goanpeca commented on spyder/plugins/plots/plugin.py:
```diff
@@ -47,7 +47,8 @@ def __init__(self, parent):
     def get_settings(self):
```
@ccordoba12 do you think it is worth having a method like this one as part of the plugin api?

ccordoba12 replied:
I don't think so.

ccordoba12 commented:
@gepcel, tests are failing for Python 2.

ccordoba12 commented on spyder/plugins/plots/widgets/figurebrowser.py:
```diff
@@ -481,8 +481,8 @@ def zoom_out(self):
         """Scale the image size."""
-        fwidth = self.figcanvas.fwidth
-        fheight = self.figcanvas.fheight
+        fwidth = self.figcanvas.fwidth * 1.0
```
I think it should be more elegant to make use here `float` instead of multiplying by `1.0`. Also for all changes you made in this commit.

jnsebgosselin replied:
@ccordoba12 Can we add `from __future__ import division` in this module instead to solve this for Python 2?

ccordoba12 replied:
Oh yeah, that's better! @gepcel, please do that instead.

gepcel replied:
Good to learn about `from __future__ import division`

goanpeca commented:
@ccordoba12 good to go?

ccordoba12 commented:
We're waiting for @jnsebgosselin's review.

goanpeca commented:
@jnsebgosselin :-p ?

jnsebgosselin commented:
@goanpeca I've tested it locally and I agree with the way this feature is implementer as it is. I've also reviewed the code and everything LGTM. I apologize for taking so long to review this and thanks for waiting, this is much appreciated.

@gepcel Thanks for the awesome work on this, this is going to be very handy!

ccordoba12 commented:
Thanks @jnsebgosselin for the review! Merging then.
